libraries/ts-command-line/src/CommandLineParameterProvider.ts

```diff
@@ -2,56 +2,77 @@
 // See LICENSE in the project root for license information.
 
 import * as argparse from 'argparse';
-import * as colors from 'colors';
 import {
-  IBaseCommandLineDefinition,
   ICommandLineFlagDefinition,
   ICommandLineStringDefinition,
   ICommandLineStringListDefinition,
   ICommandLineIntegerDefinition,
-  ICommandLineOptionDefinition
+  ICommandLineChoiceDefinition
 } from './CommandLineDefinition';
 
 import {
   CommandLineParameter,
-  ICommandLineParserData,
-  IConverterFunction,
+  CommandLineParameterWithArgument,
   CommandLineFlagParameter,
   CommandLineStringParameter,
   CommandLineStringListParameter,
   CommandLineIntegerParameter,
-  CommandLineOptionParameter
+  CommandLineChoiceParameter,
+  CommandLineParameterKind
 } from './CommandLineParameter';
 
+// (This source file uses "any" in many places for abstract parameter values)
+// tslint:disable:no-any
+
+/**
+ * This is the argparse result data object
+ * @internal
+ */
+export interface ICommandLineParserData {
+  action: string;
+  [key: string]: any;
+}
+
 /**
  * This is the common base class for CommandLineAction and CommandLineParser
  * that provides functionality for defining command-line parameters.
  *
  * @public
  */
-abstract class CommandLineParameterProvider {
+export abstract class CommandLineParameterProvider {
   private static _keyCounter: number = 0;
 
-  /**
-   * NOTE: THIS IS INTERNAL.  IN THE FUTURE, WE MAY REPLACE "argparse" WITH A DIFFERENT ENGINE.
-   * @internal
-   */
-  protected _argumentParser: argparse.ArgumentParser;
-
-  /* tslint:disable-next-line:no-any */
   private _parameters: CommandLineParameter<any>[];
-  private _keys: Map<string, string>;
+  private _parametersByLongName: Map<string, CommandLineParameter<any>>;
 
+  /** @internal */
+  // Third party code should not inherit subclasses or call this constructor
   constructor() {
     this._parameters = [];
-    this._keys = new Map<string, string>();
+    this._parametersByLongName = new Map<string, CommandLineParameter<any>>();
   }
 
   /**
-   * The child class should implement this hook to define its command-line parameters,
-   * e.g. by calling defineFlagParameter().
+   * Defines a command-line parameter whose value must be a string from a fixed set of
+   * allowable choices (similar to an enum).
+   *
+   * @remarks
```

Comment:
clarify the scaffolding scenario and the dynamic scenario

Reply:
I wrote a whole README.md

---
In reply to: [182194855](https://github.com/Microsoft/web-build-tools/pull/611#discussion_r182194855) [](ancestors = 182194855)